File: plugin/src/main/java/com/github/topi314/lavasearch/plugin/SearchRestHandler.java
package com.github.topi314.lavasearch.plugin;

import com.github.topi314.lavasearch.SearchManager;
import com.github.topi314.lavasearch.protocol.SearchResult;
import com.github.topi314.lavasearch.protocol.SearchType;
import jakarta.servlet.http.HttpServletRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Arrays;
import java.util.HashSet;
import java.util.stream.Collectors;

@RestController
public class SearchRestHandler {

	private static final Logger log = LoggerFactory.getLogger(SearchRestHandler.class);

	private final SearchManager searchManager;

	public SearchRestHandler(SearchManager searchManager) {
		this.searchManager = searchManager;
		log.info("SearchRestHandler created");
	}

	@GetMapping("/v4/loadsearch")
	public ResponseEntity<SearchResult> loadSearch(HttpServletRequest request, @RequestParam String query, @RequestParam(required = false) String rawTypes) {
		log.debug("laodSearch called with query: {}, types: {}", query, rawTypes);
		var types = rawTypes == null ? new HashSet<SearchType>() : Arrays.stream(rawTypes.split(",")).map(s -> SearchType.valueOf(s.trim())).collect(Collectors.toSet());
		var result = searchManager.loadSearch(query, types);
		if (result != null) {
			return ResponseEntity.ok(result);
		}
		return ResponseEntity.notFound().build();
	}
}

File: plugin/src/main/java/com/github/topi314/lavasearch/plugin/SearchRestHandler.kt
package com.github.topi314.lavasearch.plugin

import com.github.topi314.lavasearch.AudioSearchResult
import com.github.topi314.lavasearch.SearchManager
import com.github.topi314.lavasearch.api.SearchResultPluginInfoModifier
import com.github.topi314.lavasearch.protocol.SearchResult
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager
import dev.arbjerg.lavalink.api.AudioPluginInfoModifier
import jakarta.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.util.*

@RestController
class SearchRestHandler(
    private val searchManager: SearchManager,
    private val playerManager: AudioPlayerManager,
    private val pluginInfoModifiers: List<AudioPluginInfoModifier>,
    private val searchResultInfoModifiers: List<SearchResultPluginInfoModifier>
) {

    companion object {
        private val log = LoggerFactory.getLogger(SearchRestHandler::class.java)
    }

    @GetMapping("/v4/loadsearch")
    fun loadSearch(
        request: HttpServletRequest?,
        @RequestParam query: String?,
        @RequestParam(required = false) types: String?
    ): ResponseEntity<SearchResult> {
        log.debug("loadSearch called with query: {}, types: {}", query, types)
        val finalTypes = if (types.isNullOrBlank()) emptySet() else types.split(""",\s*""".toRegex()).mapNotNull { AudioSearchResult.Type.fromName(it) }.toSet()
        val result = searchManager.loadSearch(query, finalTypes)
        return if (result != null) {
            ResponseEntity.ok(result.toSearchResult(playerManager, pluginInfoModifiers, searchResultInfoModifiers))
        } else ResponseEntity.notFound().build()
    }
}
